x-pack/legacy/plugins/apm/public/components/app/Settings/CustomizeUI/CustomActionsOverview/CustomActionsFlyout/index.tsx

```diff
@@ -19,91 +14,133 @@ import {
 } from '@elastic/eui';
 import { i18n } from '@kbn/i18n';
 import React, { useState } from 'react';
-import { SettingsSection } from './SettingsSection';
-import { ServiceForm } from '../../../../../shared/ServiceForm';
+import { Controller, useForm } from 'react-hook-form';
+import { isEmpty } from 'lodash';
+import { CustomAction } from '../../../../../../../../../../plugins/apm/server/lib/settings/custom_action/custom_action_types';
+import { useApmPluginContext } from '../../../../../../hooks/useApmPluginContext';
+import { useCallApmApi } from '../../../../../../hooks/useCallApmApi';
+import { ActionSection } from './ActionSection';
+import { Filter, FiltersSection } from './FiltersSection';
+import { FlyoutFooter } from './FlyoutFooter';
+import { saveCustomAction } from './saveCustomAction';
 
 interface Props {
   onClose: () => void;
+  customActionSelected?: CustomAction;
+  onSave: () => void;
+  onDelete: () => void;
 }
 
-export const CustomActionsFlyout = ({ onClose }: Props) => {
-  const [serviceName, setServiceName] = useState('');
-  const [environment, setEnvironment] = useState('');
-  const [label, setLabel] = useState('');
-  const [url, setURL] = useState('');
-  return (
-    <EuiPortal>
-      <EuiFlyout ownFocus onClose={onClose} size="s">
-        <EuiFlyoutHeader hasBorder>
-          <EuiTitle size="s">
-            <h2>
-              {i18n.translate(
-                'xpack.apm.settings.customizeUI.customActions.flyout.title',
-                {
-                  defaultMessage: 'Create custom action'
-                }
-              )}
-            </h2>
-          </EuiTitle>
-        </EuiFlyoutHeader>
-        <EuiFlyoutBody>
-          <EuiText>
-            <p>
-              {i18n.translate(
-                'xpack.apm.settings.customizeUI.customActions.flyout.label',
-                {
-                  defaultMessage:
-                    "This action will be shown in the 'Actions' context menu for the trace and error detail components. You can specify any number of links, but only the first three will be shown, in alphabetical order."
-                }
-              )}
-            </p>
-          </EuiText>
-          <EuiSpacer size="l" />
-          <ServiceForm
-            isReadOnly={false}
-            serviceName={serviceName}
-            onServiceNameChange={setServiceName}
-            environment={environment}
-            onEnvironmentChange={setEnvironment}
-          />
+export interface CustomActionFormData extends Omit<CustomAction, 'filters'> {
+  filters: Filter[];
+}
 
-          <EuiSpacer size="l" />
+const convertFiltersToArray = (filters?: CustomAction['filters']) => {
+  if (filters && Object.keys(filters).length) {
+    return Object.keys(filters).map(key => {
+      return { key, value: filters[key] || '' };
+    });
+  }
+};
```

Comment:
Would be great if we could avoid converting back and forth between data structures. If we can't avoid it this can at least be simplified a bit using a default param and `object.entries`:
```js
const convertFiltersToArray = (filters = {}) => {
  return Object.entries(filters).map(([key, value]) => ({ key, value: value || '' }));
};
```

Reply:
I'm converting it to array because it's simpler to work with it when adding, editing and removing an action since I can do it only passing the index of the element that has to be modified. But I liked what you suggested. Instead of working with an array of object I'm going to work with an array of array, using Object.entries and Object.fromEntries. WDYT?

Reply:
@sqren WDYT about 👆🏻